### PR TITLE
fix(node): resolve types via package.json for directory import

### PR DESCRIPTION
**Pull request by dsherret**
Does a package resolve when resolving types for a directory (copying the behaviour that typescript does).

**Please merge on approval**

### CODE REVIEW

File: tests/specs/npm/check_pkg_json_import/main.out
Comment (dsherret) [inline tests/specs/npm/check_pkg_json_import/main.out:1]: This is the current incorrect output.

File: ext/node/resolution.rs [hunk 1]
```diff
@@ -502,29 +517,49 @@ impl NodeResolver {
...
+            NodeModuleKind::Esm => DEFAULT_CONDITIONS,
+            NodeModuleKind::Cjs => REQUIRE_CONDITIONS,
+          },
+          NodeResolutionMode::Types,
+          permissions,
+        )?;
+        if let Some(resolution) = maybe_resolution {
+          return Ok(Some(resolution));
+        }
+      }
```
Comment (dsherret) [inline ext/node/resolution.rs:547]: I looked into it and this seems to be done first before the index.d.ts probing.

File: ext/node/polyfills/01_require.js [hunk 2]
```diff
@@ -491,31 +491,6 @@ Module.globalPaths = modulePaths;
...
-  const filename = resolved;
-  const actual = tryFile(filename, false);
-  if (actual) {
-    return actual;
-  }
-  throw new ERR_MODULE_NOT_FOUND(
-    filename,
-    path.resolve(pkgPath, "package.json"),
-  );
-}
```
Comment (bartlomieju) [inline ext/node/polyfills/01_require.js:518]: IIRC this was quite an important bit and some packages were checking for this `ERR_MODULE_NOT_FOUND` `code` property. Does this value still match after this change?

Comment (dsherret) [reply ext/node/polyfills/01_require.js:518]: This code was not used so I removed it.

Comment (bartlomieju) [reply ext/node/polyfills/01_require.js:518]: Fair enough, let's land as is, but just a warning for the future. I'm not 100% sure this was the exact code path that was checking for this code.

**Review by bartlomieju — APPROVED** [review]

**Pull request merged**